doc/dev/kclient.rst

```diff
@@ -0,0 +1,345 @@
+Testing changes to the Linux Kernel CephFS driver
+=================================================
+
+This walkthrough will explain one (opinionated) way to do testing of the Linux
+kernel client against a development cluster. We will try to mimimize any
+assumptions about pre-existing knowledge of how to do kernel builds or any
+related best-practices.
+
+Step One: build the kernel
+==========================
+
+Configure the kernel:
+
+.. bash::
+
+    make defconfig
+
+
+We now have a kernel config with reasonable defaults for the architecture you're
+building on. The next thing to do is to enable configs which will build Ceph and/or
+provide functionality we need to do testing.
+
+.. bash::
+
+    cat > ~/.ceph.config <<EOF
+    CONFIG_XFS_FS=y
+    CONFIG_CEPH_FS=y
+    CONFIG_CEPH_FSCACHE=y
+    CONFIG_CEPH_FS_POSIX_ACL=y
+    CONFIG_CEPH_FS_SECURITY_LABEL=y
+    CONFIG_CEPH_LIB_PRETTYDEBUG=y
+    CONFIG_DYNAMIC_DEBUG=y
+    CONFIG_DYNAMIC_DEBUG_CORE=y
+    CONFIG_KGDB=y
+    CONFIG_FRAME_POINTER=y
+    CONFIG_KGDB_SERIAL_CONSOLE=y
+    EOF
+
+Note, beyond enabling Ceph-related configs, we are also enabling some useful
+debug configs and XFS (as an alternative to ext4 if needed for our root file
+system).
+
+Now, merge the configs.
+
+
+.. bash::
+
+
+    scripts/kconfig/merge_config.sh .config ~/.ceph.config
+
+
+Finally, build the kernel:
+
+.. bash::
+
+    make -j
```

Comment:
The detail steps will be:

```
1, git clone https://github.com/ceph/ceph-client.git ; cd ceph-client; git checkout -b my-testing origin/testing
2, generated the '.config' file under the ceph-client/ dir.
3, make menuconfig # do nothing and just close it
4, make -j
5, make modules_install; make install
6, change the grubenv
7, reboot
```

The 'origin/testing' branch will be the develop branch and will include all the latest changes in the upstream.

Reply:
`make install` should be unnecessary with a VM but I will add the `git clone` bits.

Reply:
Pardon, it should be: `sudo make modules_install; sudo make install`.

Usually the `sudo make install` is needed when you install the upstream kernel for the first time and later whenever after you sync the local code from the remote repo.

Reply:
And when you are trying to code the ceph code only then you can just replace the `ceph.ko` or/and `libceph.ko` instead of running the `sudo make modules_install` every time.

Reply:
> 3, make menuconfig # do nothing and just close it

You need to select `Ceph distributed file system` for `make -j` or `make modules` to build it.

Reply:
@lxbsz sure but wouldn't it be easier to just *not* build modules? If you use `=y` for most/all configs (especially Ceph ones) then you don't need to install anything. Just use the bzImage as your VM's kernel, done.